faucet/src/faucet.rs

```diff
@@ -202,6 +244,225 @@ pub fn faucet_error(source: CapeWalletError) -> tide::Error {
+    fn grant(&mut self, key: UserPubKey, max_grants: usize) -> Result<bool, FaucetError> {
+        let grants_given = self.index[&key] + 1;
+        if grants_given >= max_grants {
+            // If this is the last grant to this key, remove it from the index.
+            self.remove(&key)?;
+            Ok(false)
+        } else {
+            // Update the entry in our persistent log.
+            self.queue
+                .store_resource(&(key.clone(), Some(grants_given)))?;
+            self.queue.commit_version().unwrap();
+            self.store.commit_version().unwrap();
+            // If successful, update our in-memory index.
+            self.index.insert(key, grants_given);
+            Ok(true)
+        }
+    }
+
+    /// Remove an element from the persistent set.
+    fn remove(&mut self, key: &UserPubKey) -> Result<(), FaucetError> {
+        // Make a persistent note to remove the key.
+        self.queue.store_resource(&(key.clone(), None))?;
+        self.queue.commit_version().unwrap();
+        self.store.commit_version().unwrap();
+        // Update our in-memory set.
+        self.index.remove(key);
+        Ok(())
+    }
+}
+
+impl FaucetQueue {
+    async fn load(store: &Path, max_len: Option<usize>) -> Result<Self, FaucetError> {
+        // Load from storage.
+        let mut loader = AtomicStoreLoader::load(store, "queue")?;
+        let persistent_queue = AppendLog::load(&mut loader, Default::default(), "requests", 1024)?;
+        let store = AtomicStore::open(loader)?;
+
+        // Traverse the persisted queue entries backwards. This ensures that we encounter the most
+        // recent value for each key first. If the most recent value for a given key is `Some(n)`,
+        // it gets added to the index. If it is `None`, we just store `None` in `index` so that if
+        // we see this key again, we know we are not seeing the most recent value.
+        let mut index = HashMap::new();
+        // In addition, for the most recent `Some(0)` entry for each `key`, we also add that key to
+        // the message channel, as long as there is not a more recent `None` entry. We use the set
+        // `processed` to keep track of which elements have already been processed into the message
+        // channel if necessary. An element is `processed` if we have added it to the message
+        // channel, or if we have encountered a `None` entry for it and skipped it.
+        let mut processed = HashSet::new();
+        // We are encountering requests in reverse order, so if we need to add them to the queue, we
+        // will add them to this [Vec] and then reverse it at the end before adding them to the
+        // message channel.
+        let mut queue = Vec::new();
+        let entries: Vec<(UserPubKey, Option<usize>)> =
+            persistent_queue.iter().collect::<Result<_, _>>()?;
+        for (key, val) in entries.into_iter().rev() {
+            if !index.contains_key(&key) {
```

Comment:
I think this is the same as `index.insert(key.clone(), val);`, so we could also write it in the more compact version.